<?xml version="1.0" encoding="UTF-8"?>
<ui version="4.0">
 <class>MainWindow</class>
 <widget class="QMainWindow" name="MainWindow">
  <property name="geometry">
   <rect>
    <x>0</x>
    <y>0</y>
    <width>1004</width>
    <height>668</height>
   </rect>
  </property>
  <property name="sizePolicy">
   <sizepolicy hsizetype="Preferred" vsizetype="Preferred">
    <horstretch>0</horstretch>
    <verstretch>0</verstretch>
   </sizepolicy>
  </property>
  <property name="minimumSize">
   <size>
    <width>1004</width>
    <height>668</height>
   </size>
  </property>
  <property name="maximumSize">
   <size>
    <width>1004</width>
    <height>668</height>
   </size>
  </property>
  <property name="windowTitle">
   <string>拓扑控制器 wdLin @USTC 2015</string>
  </property>
  <widget class="QWidget" name="centralWidget">
   <widget class="QGroupBox" name="serialInfoGroupBox">
    <property name="geometry">
     <rect>
      <x>10</x>
      <y>20</y>
      <width>251</width>
      <height>91</height>
     </rect>
    </property>
    <property name="title">
     <string>串口信息</string>
    </property>
    <property name="flat">
     <bool>false</bool>
    </property>
    <widget class="QPushButton" name="onOffSerialPortButton">
     <property name="geometry">
      <rect>
       <x>30</x>
       <y>60</y>
       <width>75</width>
       <height>23</height>
      </rect>
     </property>
     <property name="text">
      <string>打开串口</string>
     </property>
    </widget>
    <widget class="QLabel" name="showSerialPortLabel">
     <property name="geometry">
      <rect>
       <x>130</x>
       <y>60</y>
       <width>81</width>
       <height>20</height>
      </rect>
     </property>
     <property name="text">
      <string>串口没有打开！</string>
     </property>
    </widget>
    <widget class="QComboBox" name="serialPortComboBox">
     <property name="geometry">
      <rect>
       <x>60</x>
       <y>20</y>
       <width>61</width>
       <height>21</height>
      </rect>
     </property>
     <item>
      <property name="text">
       <string>COM1</string>
      </property>
     </item>
     <item>
      <property name="text">
       <string>COM2</string>
      </property>
     </item>
     <item>
      <property name="text">
       <string>COM3</string>
      </property>
     </item>
     <item>
      <property name="text">
       <string>COM4</string>
      </property>
     </item>
     <item>
      <property name="text">
       <string>COM5</string>
      </property>
     </item>
     <item>
      <property name="text">
       <string>COM6</string>
      </property>
     </item>
     <item>
      <property name="text">
       <string>COM7</string>
      </property>
     </item>
     <item>
      <property name="text">
       <string>COM8</string>
      </property>
     </item>
     <item>
      <property name="text">
       <string>COM9</string>
      </property>
     </item>
    </widget>
    <widget class="QComboBox" name="baudRateComboBox">
     <property name="geometry">
      <rect>
       <x>180</x>
       <y>20</y>
       <width>61</width>
       <height>22</height>
      </rect>
     </property>
     <property name="currentIndex">
      <number>3</number>
     </property>
     <item>
      <property name="text">
       <string>1200</string>
      </property>
     </item>
     <item>
      <property name="text">
       <string>2400</string>
      </property>
     </item>
     <item>
      <property name="text">
       <string>4800</string>
      </property>
     </item>
     <item>
      <property name="text">
       <string>9600</string>
      </property>
     </item>
     <item>
      <property name="text">
       <string>19200</string>
      </property>
     </item>
     <item>
      <property name="text">
       <string>38400</string>
      </property>
     </item>
     <item>
      <property name="text">
       <string>57600</string>
      </property>
     </item>
     <item>
      <property name="text">
       <string>115200</string>
      </property>
     </item>
    </widget>
    <widget class="QLabel" name="staticSerialPortLabel">
     <property name="geometry">
      <rect>
       <x>10</x>
       <y>20</y>
       <width>54</width>
       <height>21</height>
      </rect>
     </property>
     <property name="text">
      <string>串口号：</string>
     </property>
    </widget>
    <widget class="QLabel" name="staticBaudRateLabel">
     <property name="geometry">
      <rect>
       <x>130</x>
       <y>20</y>
       <width>54</width>
       <height>21</height>
      </rect>
     </property>
     <property name="text">
      <string>波特率：</string>
     </property>
    </widget>
   </widget>
   <widget class="QGroupBox" name="switchInfoGroupBox">
    <property name="geometry">
     <rect>
      <x>290</x>
      <y>10</y>
      <width>701</width>
      <height>191</height>
     </rect>
    </property>
    <property name="title">
     <string>交换机信息</string>
    </property>
    <widget class="QGroupBox" name="addSwitchInfoGroupBox">
     <property name="geometry">
      <rect>
       <x>10</x>
       <y>20</y>
       <width>371</width>
       <height>111</height>
      </rect>
     </property>
     <property name="title">
      <string>增加交换机</string>
     </property>
     <widget class="QLabel" name="staticSwitchNameLabel">
      <property name="geometry">
       <rect>
        <x>10</x>
        <y>20</y>
        <width>54</width>
        <height>21</height>
       </rect>
      </property>
      <property name="text">
       <string>名字：</string>
      </property>
     </widget>
     <widget class="QLabel" name="staticSwitchPortLabel">
      <property name="geometry">
       <rect>
        <x>10</x>
        <y>50</y>
        <width>54</width>
        <height>21</height>
       </rect>
      </property>
      <property name="text">
       <string>端口：</string>
      </property>
     </widget>
     <widget class="QLabel" name="staticSwitchXLabel">
      <property name="geometry">
       <rect>
        <x>160</x>
        <y>20</y>
        <width>61</width>
        <height>16</height>
       </rect>
      </property>
      <property name="text">
       <string>X坐标(km)：</string>
      </property>
     </widget>
     <widget class="QLabel" name="staticSwitchYLabel">
      <property name="geometry">
       <rect>
        <x>160</x>
        <y>50</y>
        <width>61</width>
        <height>21</height>
       </rect>
      </property>
      <property name="text">
       <string>Y坐标(km)：</string>
      </property>
     </widget>
     <widget class="QLineEdit" name="switchNameLineEdit">
      <property name="geometry">
       <rect>
        <x>50</x>
        <y>20</y>
        <width>91</width>
        <height>20</height>
       </rect>
      </property>
     </widget>
     <widget class="QComboBox" name="switchPortComboBox">
      <property name="geometry">
       <rect>
        <x>60</x>
        <y>50</y>
        <width>51</width>
        <height>22</height>
       </rect>
      </property>
      <item>
       <property name="text">
        <string>1</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>2</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>3</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>4</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>5</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>6</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>7</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>8</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>9</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>10</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>11</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>12</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>13</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>14</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>15</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>16</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>17</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>18</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>19</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>20</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>21</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>22</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>23</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>24</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>25</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>26</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>27</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>28</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>29</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>30</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>31</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>32</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>33</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>34</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>35</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>36</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>37</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>38</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>39</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>40</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>41</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>42</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>43</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>44</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>45</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>46</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>47</string>
       </property>
      </item>
      <item>
       <property name="text">
        <string>48</string>
       </property>
      </item>
     </widget>
     <widget class="QLineEdit" name="switchXLineEdit">
      <property name="geometry">
       <rect>
        <x>230</x>
        <y>20</y>
        <width>101</width>
        <height>20</height>
       </rect>
      </property>
     </widget>
     <widget class="QLineEdit" name="switchYLineEdit">
      <property name="geometry">
       <rect>
        <x>230</x>
        <y>50</y>
        <width>101</width>
        <height>20</height>
       </rect>
      </property>
     </widget>
     <widget class="QPushButton" name="addSwitchButton">
      <property name="geometry">
       <rect>
        <x>110</x>
        <y>80</y>
        <width>141</width>
        <height>23</height>
       </rect>
      </property>
      <property name="text">
       <string>添加</string>
      </property>
     </widget>
    </widget>
    <widget class="QGroupBox" name="distanceInfoGroupBox">
     <property name="geometry">
      <rect>
       <x>10</x>
       <y>140</y>
       <width>371</width>
       <height>41</height>
      </rect>
     </property>
     <property name="title">
      <string/>
     </property>
     <widget class="QLabel" name="staticConnectedDistanceLabel">
      <property name="geometry">
       <rect>
        <x>10</x>
        <y>10</y>
        <width>81</width>
        <height>21</height>
       </rect>
      </property>
      <property name="text">
       <string>连接距离(km)：</string>
      </property>
     </widget>
     <widget class="QLineEdit" name="connectedDistanceLineEdit">
      <property name="geometry">
       <rect>
        <x>90</x>
        <y>10</y>
        <width>81</width>
        <height>21</height>
       </rect>
      </property>
     </widget>
     <widget class="QPushButton" name="setConnectedDistanceButton">
      <property name="geometry">
       <rect>
        <x>180</x>
        <y>10</y>
        <width>101</width>
        <height>23</height>
       </rect>
      </property>
      <property name="text">
       <string>设置连接距离</string>
      </property>
     </widget>
     <widget class="QLabel" name="staticCurrentDistanceLabel">
      <property name="geometry">
       <rect>
        <x>290</x>
        <y>0</y>
        <width>81</width>
        <height>20</height>
       </rect>
      </property>
      <property name="text">
       <string>当前距离(km)：</string>
      </property>
     </widget>
     <widget class="QLabel" name="showCurrentDistanceLabel">
      <property name="geometry">
       <rect>
        <x>300</x>
        <y>20</y>
        <width>54</width>
        <height>12</height>
       </rect>
      </property>
      <property name="text">
       <string>30</string>
      </property>
      <property name="alignment">
       <set>Qt::AlignCenter</set>
      </property>
     </widget>
    </widget>
    <widget class="QGroupBox" name="switchListInfoGroupBox">
     <property name="geometry">
      <rect>
       <x>400</x>
       <y>10</y>
       <width>291</width>
       <height>171</height>
      </rect>
     </property>
     <property name="title">
      <string/>
     </property>
     <widget class="QListWidget" name="switchListWidget">
      <property name="geometry">
       <rect>
        <x>10</x>
        <y>10</y>
        <width>271</width>
        <height>121</height>
       </rect>
      </property>
     </widget>
     <widget class="QPushButton" name="modifySwitchButton">
      <property name="geometry">
       <rect>
        <x>50</x>
        <y>140</y>
        <width>75</width>
        <height>23</height>
       </rect>
      </property>
      <property name="text">
       <string>修改</string>
      </property>
     </widget>
     <widget class="QPushButton" name="deleteSwitchButton">
      <property name="geometry">
       <rect>
        <x>150</x>
        <y>140</y>
        <width>75</width>
        <height>23</height>
       </rect>
      </property>
      <property name="text">
       <string>删除</string>
      </property>
     </widget>
    </widget>
   </widget>
   <widget class="PaintWidget" name="paintWidget" native="true">
    <property name="geometry">
     <rect>
      <x>280</x>
      <y>220</y>
      <width>710</width>
      <height>410</height>
     </rect>
    </property>
   </widget>
   <widget class="QGroupBox" name="areaInfoGroupBox">
    <property name="geometry">
     <rect>
      <x>10</x>
      <y>130</y>
      <width>251</width>
      <height>181</height>
     </rect>
    </property>
    <property name="title">
     <string>区域信息</string>
    </property>
    <widget class="QLabel" name="staticMaxXLabel">
     <property name="geometry">
      <rect>
       <x>20</x>
       <y>20</y>
       <width>54</width>
       <height>21</height>
      </rect>
     </property>
     <property name="text">
      <string>MaxX(km)：</string>
     </property>
    </widget>
    <widget class="QLabel" name="staticMaxYLabel">
     <property name="geometry">
      <rect>
       <x>20</x>
       <y>50</y>
       <width>54</width>
       <height>16</height>
      </rect>
     </property>
     <property name="text">
      <string>MaxY(km)：</string>
     </property>
    </widget>
    <widget class="QLabel" name="staticSpeedLabel">
     <property name="geometry">
      <rect>
       <x>10</x>
       <y>80</y>
       <width>91</width>
       <height>21</height>
      </rect>
     </property>
     <property name="text">
      <string>速度(km/h)：</string>
     </property>
    </widget>
    <widget class="QLineEdit" name="maxXLineEdit">
     <property name="geometry">
      <rect>
       <x>80</x>
       <y>20</y>
       <width>61</width>
       <height>20</height>
      </rect>
     </property>
    </widget>
    <widget class="QLineEdit" name="maxYLineEdit">
     <property name="geometry">
      <rect>
       <x>80</x>
       <y>50</y>
       <width>61</width>
       <height>20</height>
      </rect>
     </property>
    </widget>
    <widget class="QLineEdit" name="speedLineEdit">
     <property name="geometry">
      <rect>
       <x>80</x>
       <y>80</y>
       <width>61</width>
       <height>20</height>
      </rect>
     </property>
    </widget>
    <widget class="QPushButton" name="setMaxXButton">
     <property name="geometry">
      <rect>
       <x>160</x>
       <y>20</y>
       <width>75</width>
       <height>23</height>
      </rect>
     </property>
     <property name="text">
      <string>设置MaxX</string>
     </property>
    </widget>
    <widget class="QPushButton" name="setMaxYButton">
     <property name="geometry">
      <rect>
       <x>160</x>
       <y>50</y>
       <width>75</width>
       <height>23</height>
      </rect>
     </property>
     <property name="text">
      <string>设置MaxY</string>
     </property>
    </widget>
    <widget class="QPushButton" name="setSpeedButton">
     <property name="geometry">
      <rect>
       <x>160</x>
       <y>80</y>
       <width>75</width>
       <height>23</height>
      </rect>
     </property>
     <property name="text">
      <string>设置速度</string>
     </property>
    </widget>
    <widget class="QLabel" name="staticCurrentMaxXLabel">
     <property name="geometry">
      <rect>
       <x>30</x>
       <y>110</y>
       <width>91</width>
       <height>20</height>
      </rect>
     </property>
     <property name="text">
      <string>当前MaxX(km)：</string>
     </property>
    </widget>
    <widget class="QLabel" name="staticCurrentMaxYLabel">
     <property name="geometry">
      <rect>
       <x>30</x>
       <y>130</y>
       <width>91</width>
       <height>20</height>
      </rect>
     </property>
     <property name="text">
      <string>当前MaxY(km)：</string>
     </property>
    </widget>
    <widget class="QLabel" name="staticCurrentSpeedLabel">
     <property name="geometry">
      <rect>
       <x>30</x>
       <y>150</y>
       <width>91</width>
       <height>20</height>
      </rect>
     </property>
     <property name="text">
      <string>当前速度(km/h)：</string>
     </property>
    </widget>
    <widget class="QLabel" name="showCurrentMaxXLabel">
     <property name="geometry">
      <rect>
       <x>130</x>
       <y>110</y>
       <width>54</width>
       <height>12</height>
      </rect>
     </property>
     <property name="text">
      <string>0</string>
     </property>
     <property name="alignment">
      <set>Qt::AlignCenter</set>
     </property>
    </widget>
    <widget class="QLabel" name="showCurrentMaxYLabel">
     <property name="geometry">
      <rect>
       <x>130</x>
       <y>130</y>
       <width>54</width>
       <height>12</height>
      </rect>
     </property>
     <property name="text">
      <string>0</string>
     </property>
     <property name="alignment">
      <set>Qt::AlignCenter</set>
     </property>
    </widget>
    <widget class="QLabel" name="showCurrentSpeedLabel">
     <property name="geometry">
      <rect>
       <x>130</x>
       <y>150</y>
       <width>54</width>
       <height>16</height>
      </rect>
     </property>
     <property name="text">
      <string>0</string>
     </property>
     <property name="alignment">
      <set>Qt::AlignCenter</set>
     </property>
    </widget>
   </widget>
   <widget class="QGroupBox" name="runInfoGroupBox">
    <property name="geometry">
     <rect>
      <x>10</x>
      <y>330</y>
      <width>251</width>
      <height>311</height>
     </rect>
    </property>
    <property name="title">
     <string>运行</string>
    </property>
    <widget class="QPushButton" name="onceConfigureButton">
     <property name="geometry">
      <rect>
       <x>60</x>
       <y>10</y>
       <width>111</width>
       <height>23</height>
      </rect>
     </property>
     <property name="text">
      <string>静态配置</string>
     </property>
    </widget>
    <widget class="QPushButton" name="dynamicConfigureButton">
     <property name="geometry">
      <rect>
       <x>60</x>
       <y>70</y>
       <width>111</width>
       <height>23</height>
      </rect>
     </property>
     <property name="text">
      <string>开始动态配置</string>
     </property>
    </widget>
    <widget class="QPushButton" name="runSwithButton">
     <property name="geometry">
      <rect>
       <x>60</x>
       <y>130</y>
       <width>111</width>
       <height>23</height>
      </rect>
     </property>
     <property name="text">
      <string>开始运动</string>
     </property>
    </widget>
    <widget class="QLabel" name="staticOnceConfigureLabel">
     <property name="geometry">
      <rect>
       <x>20</x>
       <y>40</y>
       <width>201</width>
       <height>20</height>
      </rect>
     </property>
     <property name="text">
      <string>根据当前状态，静态配置一次。</string>
     </property>
     <property name="alignment">
      <set>Qt::AlignCenter</set>
     </property>
    </widget>
    <widget class="QLabel" name="staticDynamicConfigureLabel">
     <property name="geometry">
      <rect>
       <x>30</x>
       <y>100</y>
       <width>181</width>
       <height>20</height>
      </rect>
     </property>
     <property name="text">
      <string>根据当前状态，动态配置拓扑。</string>
     </property>
     <property name="alignment">
      <set>Qt::AlignCenter</set>
     </property>
    </widget>
    <widget class="QLabel" name="staticRunLabel">
     <property name="geometry">
      <rect>
       <x>30</x>
       <y>160</y>
       <width>171</width>
       <height>20</height>
      </rect>
     </property>
     <property name="text">
      <string>根据目前设置，运动交换机。</string>
     </property>
     <property name="alignment">
      <set>Qt::AlignCenter</set>
     </property>
    </widget>
    <widget class="QGroupBox" name="runningInfoGroupBox">
     <property name="geometry">
      <rect>
       <x>0</x>
       <y>190</y>
       <width>251</width>
       <height>111</height>
      </rect>
     </property>
     <property name="title">
      <string/>
     </property>
     <widget class="QLabel" name="staticOnceConfigureStatusLabel">
      <property name="geometry">
       <rect>
        <x>10</x>
        <y>10</y>
        <width>61</width>
        <height>16</height>
       </rect>
      </property>
      <property name="text">
       <string>静态配置：</string>
      </property>
     </widget>
     <widget class="QLabel" name="showOnceConfigureStatusLabel">
      <property name="geometry">
       <rect>
        <x>70</x>
        <y>10</y>
        <width>181</width>
        <height>20</height>
       </rect>
      </property>
      <property name="text">
       <string>无</string>
      </property>
      <property name="alignment">
       <set>Qt::AlignCenter</set>
      </property>
     </widget>
     <widget class="QLabel" name="staticDynamicConfigureStatusLabel">
      <property name="geometry">
       <rect>
        <x>10</x>
        <y>40</y>
        <width>54</width>
        <height>21</height>
       </rect>
      </property>
      <property name="text">
       <string>动态配置：</string>
      </property>
     </widget>
     <widget class="QLabel" name="showDynamicConfigureStatusLabel">
      <property name="geometry">
       <rect>
        <x>70</x>
        <y>40</y>
        <width>181</width>
        <height>20</height>
       </rect>
      </property>
      <property name="text">
       <string>无</string>
      </property>
      <property name="alignment">
       <set>Qt::AlignCenter</set>
      </property>
     </widget>
     <widget class="QLabel" name="staticRunStatusLabel">
      <property name="geometry">
       <rect>
        <x>10</x>
        <y>70</y>
        <width>61</width>
        <height>21</height>
       </rect>
      </property>
      <property name="text">
       <string>运动状态：</string>
      </property>
     </widget>
     <widget class="QLabel" name="showRunStatusLabel">
      <property name="geometry">
       <rect>
        <x>63</x>
        <y>70</y>
        <width>191</width>
        <height>20</height>
       </rect>
      </property>
      <property name="text">
       <string>无</string>
      </property>
      <property name="alignment">
       <set>Qt::AlignCenter</set>
      </property>
     </widget>
    </widget>
   </widget>
  </widget>
  <widget class="QMenuBar" name="menuBar">
   <property name="geometry">
    <rect>
     <x>0</x>
     <y>0</y>
     <width>1004</width>
     <height>23</height>
    </rect>
   </property>
   <widget class="QMenu" name="menuSet">
    <property name="title">
     <string>文件</string>
    </property>
    <addaction name="actionSaveSetting"/>
    <addaction name="actionSaveSettingAs"/>
   </widget>
   <widget class="QMenu" name="menuGuanyu">
    <property name="title">
     <string>关于</string>
    </property>
    <addaction name="actionAbout"/>
   </widget>
   <addaction name="menuSet"/>
   <addaction name="menuGuanyu"/>
  </widget>
  <action name="actionSaveSetting">
   <property name="text">
    <string>保存配置</string>
   </property>
  </action>
  <action name="actionSaveSettingAs">
   <property name="text">
    <string>配置另存为（.csv）</string>
   </property>
  </action>
  <action name="actionAbout">
   <property name="text">
    <string>关于软件</string>
   </property>
  </action>
 </widget>
 <layoutdefault spacing="6" margin="11"/>
 <customwidgets>
  <customwidget>
   <class>PaintWidget</class>
   <extends>QWidget</extends>
   <header location="global">PaintWidget.h</header>
   <container>1</container>
  </customwidget>
 </customwidgets>
 <resources/>
 <connections/>
</ui>
